### PR TITLE
[in_app_purchase] 0.3.0 Migration guide

### PR DESCRIPTION
**Pull request by JeroenWeener**
The API changes introduced in 0.3.0 turned out to be bigger and more complicated than anticipated. This migration guide aims to help developers upgrade to 0.3.0 by giving context to the changes and treating some exemplary use cases.

The need for a migration guide was brought to my attention in flutter/flutter#127844.
A follow-up PR will update the README in the app-facing package to align with the breaking changes introduced in 0.3.0, as the current one is out-of-date.

Fixes flutter/flutter#127844

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by gmackall** [comment]
Thanks for adding this guide! I'll block some time to review this this afternoon. Requesting review from @stuartmorgan as well, as you reviewed the initial migration to billing client v5 (let me know if you'd prefer I find a different secondary reviewer Stuart 🙂 ).

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (gmackall) [inline packages/in_app_purchase/in_app_purchase_android/migration_guide.md:8]: Heads up that this now re-directs to the Android Billing Client v6 migration guide.

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (gmackall) [inline packages/in_app_purchase/in_app_purchase_android/migration_guide.md:29]: nit: Looks like there are three cases here?

**Review by gmackall — COMMENTED** [review]
This mostly LGTM, but I want to hold off on approving till I can finish reading through the linked android documentations tomorrow.

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (JeroenWeener) [reply packages/in_app_purchase/in_app_purchase_android/migration_guide.md]: Good catch. I changed the link to take one to [the section mentioning V5](https://developer.android.com/google/play/billing/migrate-gpblv6#5-or-6).

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (JeroenWeener) [reply packages/in_app_purchase/in_app_purchase_android/migration_guide.md]: You're right. Updated.

**Comment by JeroenWeener** [comment]
@gmackall sorry for the inconvenience, but I updated the subscription use cases once more based on feedback in https://github.com/flutter/flutter/issues/127844, as the previous examples missed a crucial detail in fetching subscription offer details.

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/migration_guide.md:54]: Why is the force unwrapping safe here?

Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/migration_guide.md:54]: Why is the force unwrapping safe here?

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/migration_guide.md:50]: Although it's not enforced by CI tooling for files other than README.md, we should use [code excerpts](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#readme-code) for all of the "after" examples so that we are certain that they compile, and continue to compile as the package changes. I believe once you do the setup described there it will work for any `.md` file, not just README.

(If that's not the case please let me know, and I'll update the tooling as necessary.)

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/migration_guide.md:86]: Why are these force-unwraps safe?

I think it's worth commenting these in the example code; we should be showing best practice, and best practice should generally be to only force unwrap when it's obvious (either from direct context, or an explicit comment) why it's safe.

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/migration_guide.md:122]: Same here.

**Review by stuartmorgan — COMMENTED** [review]
The content looks great! Other than converting to code excerpts, just a few minor questions.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
(Oops, I meant that to be a Request Changes for PR status tracking.)

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (JeroenWeener) [reply packages/in_app_purchase/in_app_purchase_android/migration_guide.md:52]: Thanks for pointing me to the existence of code excerpts, I like it. I was able to generate excerpts for README files, but not for the migration guide `.md` file. Could you check for me if the tooling allows for any `.md` file?

Below is (part of) the output of `dart run script/tool/bin/flutter_plugin_tools.dart update-excerpts --packages in_app_purchase`:

```
Running command: "dart run code_excerpt_updater --write-in-place --yaml --no-escape-ng-interpolation ../README.md" in /Users/jeroen/Projects/flutter-packages/packages/in_app_purchase/in_app_purchase_android/example
Processed 1 Dart/Jade/Markdown files: 0 out of 1 fragments needed updating.
Running command: "dart run code_excerpt_updater --write-in-place --yaml --no-escape-ng-interpolation README.md" in /Users/jeroen/Projects/flutter-packages/packages/in_app_purchase/in_app_purchase_android/example
Processed 1 Dart/Jade/Markdown files: 0 out of 0 fragments needed updating.
```

It looks like it is hardcoded to cater to `README.md` and `../README.md`.

Comment (stuartmorgan) [reply packages/in_app_purchase/in_app_purchase_android/migration_guide.md:52]: Oops, so it is. I tried to prescreen in the tool source, and looked for the wrong thing. I'll take a look now.

Comment (stuartmorgan) [reply packages/in_app_purchase/in_app_purchase_android/migration_guide.md:52]: https://github.com/flutter/packages/pull/4212 will fix it; once that lands you can merge/rebase. Sorry about that!

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_android/migration_guide.md:1]: This is missing `_example`; it's the example's package name rather than the plugin's package name.

File: packages/in_app_purchase/in_app_purchase_android/migration_guide.md
Comment (stuartmorgan) [reply packages/in_app_purchase/in_app_purchase_android/migration_guide.md:52]: Now it should work :)

**Review by gmackall — APPROVED** [review]
Re-adding my approval as this has gone through a couple of changes since my last review, but still LGTM!

I also was not aware of the existence of code-excerpts, but will make sure to recommend them moving forward

**Review by stuartmorgan — COMMENTED** [review]
LGTM, thanks!

**Review by stuartmorgan — APPROVED** [review]

**Comment by Hixie** [comment]
test-exempt: documentation

Does it make sense to call this file something more specific? (To put it another way, what do we plan to do when we add a second or third migration guide?)

**Comment by stuartmorgan** [comment]
> Does it make sense to call this file something more specific? (To put it another way, what do we plan to do when we add a second or third migration guide?)

If we need more migration guides we would probably put them all in the one file, with top-level sections for each major version, since we don't know what version people will be starting from.

**Comment by stuartmorgan** [comment]
Overriding stale tree status.

**Pull request merged**
